### ISSUE TITLE
Some questions about the detach of the f in the ft_netAB

### ISSUE THREAD

**Issue by Liu-1994** [NVlabs/DG-Net#40]
Hello, thank you very much for providing the implementation code of the DG-Net model.
I encountered some problems during the implementation of the project. I will be honored if you can give me some suggestions.

I found the f=f.detch() in the ft_netAB. This causes the vector f to have no gradient. Then the loss_gen_recon_f_* has no no contribution to model parameter update. Is there something wrong with me or something wrong with the code?
![1](https://user-images.githubusercontent.com/53415242/78989130-481d1400-7b65-11ea-814f-25493bb72269.png)
![2020-04-10 19-50-10屏幕截图](https://user-images.githubusercontent.com/53415242/78989140-4f442200-7b65-11ea-84a1-2989abec1ddd.png)

I will be grateful if you can give me some suggestions. Thank you！

**Comment by layumi** [comment]
Thank you @Liu-1994 .
‘f’ is the appearance code for image generation; We  do not want the generation losses to update `f`.
Thus, we use the `detach` here. In this way, `f` is mainly updated via the re-id related losses.

**Comment by Liu-1994** [comment]
@layumi Thanks for replying. I understood that. 
And I have another small question. What is the role of the `loss_gen_recon_f_a` and `loss_gen_recon_f_b` ？As `f` is the input of the generator, I think the two losses may not update `G`.

**Comment by layumi** [comment]
`Input -> Appearance Encoder(No Update) -> f (Detach)-> Decoder (Has Gradient) -> Generated Image (Has Gradient) -> Appearance Encoder(No Update, But has gradient) -> f`

**Comment by layumi** [comment]
`Input -> Appearance Encoder(No Update) -> f (Detach)-> Decoder (Has Gradient) -> Generated Image (Has Gradient) -> Appearance Encoder(No Update, But has gradient) -> f`

**Comment by Liu-1994** [comment]
@layumi  Thanks for teh replying. I have understood the process except the last step.
As the` f.detach()` is executed in the `ft_netAB.forward()`. I think the final `f` in the above process is also `detach`so there is no back propagation for f -> Appearance Encoder. When I debug the project, the `grad_fn` of `f_a` and `f_a_recon` are both None.
<img width="653" alt="1" src="https://user-images.githubusercontent.com/53415242/79034316-983cba80-7be7-11ea-91ca-af81b200e471.png">

**Comment by layumi** [comment]
Thanks for the great suggestion. 
I have not checked it yet. It may not work.
In fact, the feature reconstruction loss has a similar role with the ID recon loss.

**Comment by Liu-1994** [comment]
Thank you very much for your reply. I have nothing else to ask about this issue.
By the way, if you are convenient, could you please take a look at another issue (#39 ) I raised.

**Comment by Liu-1994** [comment]
Thank you very much for your reply. I have nothing else to ask about this issue.
By the way, if you are convenient, could you please take a look at another issue (#39 ) I raised.
